File: BabylonProject/Assets/Content/Physics/SixdofJoint.cs
using System;
using UnityEditor;
using UnityEngine;
using UnityEngine.Animations;

/**
* Editor Script Component
* @class SixdofJoint
*/
[Babylon(Class="PROJECT.SixdofJoint"), AddComponentMenu(CanvasToolsStatics.CANVAS_TOOLS_MENU + "/Content Components/Physics/Sixdof Joint")]
public class SixdofJoint : EditorScriptComponent
{
    public Rigidbody bodyA = null;
    public Rigidbody bodyB = null;

    [Auto] public Vector3 pivotA = new Vector3(0, -0.5f, 0);
    [Auto] public Vector3 pivotB = new Vector3(0, 0.5f, 0);

    [Auto] public Vector3 perpAxisA = new Vector3(1,0,0);
    [Auto] public Vector3 perpAxisB = new Vector3(1,0,0);

    [Auto] public Physics6DoFLimit[] axisLimits = new Physics6DoFLimit[] { new Physics6DoFLimit(PhysicsConstraintAxis.LINEAR_DISTANCE, 1, 2) };

    [Auto] public bool collisionsEnabled = false;
}

[Serializable]
public class Physics6DoFLimit
{
    public PhysicsConstraintAxis axis = PhysicsConstraintAxis.LINEAR_DISTANCE;
    public float minLimit = 0;
    public float maxLimit = 1;
    public float stiffness = 0;
    public float damping = 0;
    public Physics6DoFLimit(PhysicsConstraintAxis axis = PhysicsConstraintAxis.LINEAR_DISTANCE, float minLimit = 0, float maxLimit = 1, float stiffness = 0, float damping = 0)
    {
        this.axis = axis;
        this.minLimit = minLimit;
        this.maxLimit = maxLimit;
        this.stiffness = stiffness;
        this.damping = damping;
    }
}

[Serializable]
public enum PhysicsConstraintAxis
{
    LINEAR_X = 0,
    LINEAR_Y = 1,
    LINEAR_Z = 2,
    ANGULAR_X = 3,
    ANGULAR_Y = 4,
    ANGULAR_Z = 5,
    LINEAR_DISTANCE = 6
}

File: BabylonProject/Assets/Content/Physics/SixdofJoint.ts
module PROJECT {
    /**
    * Babylon Script Component
    * @class SixdofJoint
    */
    export class SixdofJoint extends BABYLON.Toolkit.ScriptComponent {
        public bodyA: BABYLON.TransformNode = null;
        public bodyB: BABYLON.TransformNode = null;
        public pivotA: BABYLON.Vector3 = new BABYLON.Vector3(0, -0.5, 0);
        public pivotB: BABYLON.Vector3 = new BABYLON.Vector3(0, 0.5, 0);
        public perpAxisA: BABYLON.Vector3 = new BABYLON.Vector3(1, 0, 0);
        public perpAxisB: BABYLON.Vector3 = new BABYLON.Vector3(1, 0, 0);
        public axisLimits: BABYLON.Physics6DoFLimit[] = null;
        public constraint: BABYLON.Physics6DoFConstraint = null;
        public collisionsEnabled: boolean = false;

        protected awake(): void {
            const bodyAInfo: BABYLON.Toolkit.IUnityTransform = this.getProperty("bodyA");
            const bodyBInfo: BABYLON.Toolkit.IUnityTransform = this.getProperty("bodyB");
            if (bodyAInfo != null && bodyBInfo != null) {
                this.bodyA = (this.transform.name === bodyAInfo.name) ? this.transform : this.getChildNode(bodyAInfo.name);
                this.bodyB = this.getChildNode(bodyBInfo.name);
                if (this.bodyA != null && this.bodyB != null) {
                    if (this.bodyA.physicsBody != null && this.bodyB.physicsBody != null) {
                        this.constraint = new BABYLON.Physics6DoFConstraint(
                            {
                                pivotA: this.pivotA,
                                pivotB: this.pivotB,
                                perpAxisA: this.perpAxisA,
                                perpAxisB: this.perpAxisB
                            },
                            this.axisLimits,
                            this.scene
                        );
                        this.bodyA.physicsBody.addConstraint(this.bodyB.physicsBody, this.constraint);
                        this.constraint.isCollisionsEnabled = this.collisionsEnabled;
                    }
                }
            }
        }

        protected destroy(): void {
            this.bodyA = null;
            this.bodyB = null;
            if (this.constraint != null) {
                this.constraint.dispose();
                this.constraint = null;
            }
        }
    }
}